---
name:
test_type: atomic

description: |
  This rule checks ...

success_criterion:
- x.x.x # (Name of Success Criterion)

test_aspects:
- # (e.g. HTTP Messages, DOM Tree, CSS Styling, Accessibility Tree, Language, etc.,)

authors:
- # (full names as found in /_data/contributors.yml@master - if not yet listed, please have authors added to the list)
---

## Test Procedure

### Applicability

The rule applies to any (??) element ...

### Expectation (1)

Each target element ...

## Assumptions

*There are currently no assumptions*

## Accessibility support

*There are no major accessibility support issues known for this rule.*

## Background

- (e.g. WCAG Techniques or links with background information mentioned in Applicability, Expectations or Assumptions)

## Test Cases

### Passed

#### Passed example 1

Description...

```html
<!-- code -->
```

#### Passed example 2

...

### Failed

#### Failed example 1

Description...

```html
<!-- code -->
```

#### Failed example 2

...

### Inapplicable

#### Inapplicable example 1

Description...

```html
<!-- code -->
```

#### Inapplicable example 2

...
