---
name:
rule_type: composite
description: |
  This rule checks ...

success_criterion:
- x.x.x # (Name of Success Criterion)

atomic_rules:
- 

authors:
- # (full names as found in `contributors` property in  `package.json` - if not yet listed, please have authors added to the list)
---

## Aggregation Definition

### Applicability

The rule applies to any (??) element ...

### Expectation

For each test target, the outcome of (at least one of / all of / any of etc.) the following rules is (passed / failed / etc.):
- (list atomic rules that are used in this composite rule)
- 

## Assumptions

*There are currently no assumptions.*

## Accessibility support

*There are no major accessibility support issues known for this rule.*

## Background

- - (e.g. WCAG Techniques or links with background information mentioned in Applicability, Expectations or Assumptions)


## Test Cases

### Passed

#### Passed example 1

Description...

```html
<!-- code -->
```

#### Passed example 2

...

### Failed

#### Failed example 1

Description...

```html
<!-- code -->
```

#### Failed example 2

...

### Inapplicable

#### Inapplicable example 1

Description...

```html
<!-- code -->
```

#### Inapplicable example 2

...
