---
id: 80f0bf
name: Video or audio has no auto-play audio.
rule_type: composite
description: |
    This rule checks that auto-play audio does not last for more than 3 seconds, or the audio has a control mechanism to stop or mute the it.
accessibility_requirements:
  wcag20:1.4.2: # Audio Control (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_rules:
  - 95e043
  - aaa1bf
authors:
  - Anne Thyme NÃ¸rregaard
  - Bryn ANderson
---

## Applicability

This rule applies to:
any HTML `<audio>`, `<video>` or `<source>` elements, with a `src` attribute referencing content with a duration of more than 3 seconds or with a `loop` attribute equal to `true`, that has an `autoplay` attribute equal to `true`, and that has either a `paused` or `muted` attribute equal to `false`.

## Expectation

For each test target, the outcome of at least one of the following rules is passed:
- Auto-play audio does not last more than 3 seconds
- Auto-play audio has a control mechanism
 
## Assumptions

*There are currently no assumptions*

## Accessibility Support

The native `<video>` and `<audio>` controls in several browser and assistive technology combinations are not keyboard accessible and the `<video>` or `<audio>` element itself may not be anounced. Authors are recomended to use custom controls for keyboard navigation and cross browser accessibility support in generakl.

## Background

- [Success Criterion 1.4.2 Audio Control](https://www.w3.org/TR/WCAG21/#audio-control)
- [Accessible Multimedia](https://developer.mozilla.org/en-US/docs/Learn/Accessibility/Multimedia)

## Test Cases

### Passed

#### Passed example 1

The `audio` element has a control mechanism.

``` html
  <audio src="../test-assets/moon-audio/moon-speech.mp3" autoplay="true" controls></audio>
```

#### Passed example 2

The `video` element does not play for longer than 3 seconds.

``` html
 <video autoplay="true">
  <source src="../test-assets/rabbit-video/video.mp4#t=8,10" type="video/mp4" />
  <source src="../test-assets/rabbit-video/video.webm#t=8,10" type="video/webm" />
</video>
```

### Failed

#### Failed example 1

The `audio` element autoplays and does not have a control mechanism.

``` html
  <audio src="../test-assets/moon-audio/moon-speech.mp3" autoplay="true"></audio>
```

#### Failed example 2

The `video` element audio autoplays for longer than 3 seconds.

``` html
 <video autoplay="true" muted="true">
  <source src="../test-assets/rabbit-video/video.mp4" type="video/mp4" />
  <source src="../test-assets/rabbit-video/video.webm" type="video/webm" />
</video>
```

### Inapplicable

#### Inapplicable example 1

The `video` element is `muted`.

``` html
 <video autoplay="true" muted="true">
  <source src="../test-assets/rabbit-video/video.mp4" type="video/mp4" />
  <source src="../test-assets/rabbit-video/video.webm" type="video/webm" />
</video>
```

#### Inapplicable example 2

The `video` element has no audio output.

``` html
 <video autoplay="true">
  <source src="../test-assets/rabbit-video/video-with-incorrect-voiceover.mp4" type="video/mp4" />
  <source src="../test-assets/rabbit-video/video-with-incorrect-voiceover.webm" type="video/webm" />
</video>
```

#### Inapplicable example 3

The `audio` element does not have an `autoplay` attribute.

``` html
  <audio src="../test-assets/moon-audio/moon-speech.mp3" controls></audio>
```
