---
name: CSS orientation lock via show/hide controls

test_type: atomic

description: |
  This rule checks that a control on the page handle the content orientation.

test_aspects:
- DOM Tree
- CSS Styling

authors:
- Audrey Maniez
- Jey Nandakumar
---

## Test Procedure

### Applicability

The rule applies to any `element` that is [visible](#visible) and [included in the accessibility tree](#included-in-the-accessibility-tree) of the [page](https://www.w3.org/TR/WCAG21/#dfn-web-page-s).

### Expectation

Each target `element` can be viewed and operated, in the user preferred orientation, where the content can be made operable either seamlessly or by activating a control in the page.

**Note:** The user preferred orientation is either the device orientation or the orientation locked by the user in the device system settings.

## Assumptions

* This test assumes that a specific display orientation is not [essential](https://www.w3.org/TR/WCAG21/#dfn-essential), which is listed as a valid exception to [SC 1.3.4]((https://www.w3.org/WAI/WCAG21/Understanding/orientation.html)). Examples where a particular display orientation may be essential are a bank check, a piano application, slides for a projector or television, or virtual reality content where binary display orientation is not applicable.

## Accessibility support

*There are no major accessibility support issues known for this rule.*

## Background

- [Understanding Success Criterion 1.3.4: Orientation](https://www.w3.org/WAI/WCAG21/Understanding/orientation.html)
- [Managing screen orientation](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model/Managing_screen_orientation)
- [MDN Orientation](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/orientation)

## Test Cases

### Passed

#### Passed example 1

A page where the content is adaptable based on orientation. In this case, a switch button can be activated to display the respective page content based on orientation.

```html
<html lang="en">
  <head>
  </head>
  <body>
    <button id='switch-button'>
      Switch Orientation
    </button>
    <section id='show-me-portrait'>
        I am operable in PORTRAIT
    </section>
    <section id='show-me-landscape'>
        I am operable in LANDSCAPE
    </section>
    <script>
      var switchButton = document.getElementById('switch-button');
      var portraitContent = document.getElementById('show-me-portrait');
      var landscapeContent = document.getElementById('show-me-landscape');
      var currentOrientation;
      function setUpOrientation(orientation) {
        if(orientation === 'portrait') {
          portraitContent.style.display = 'block';
          landscapeContent.style.display = 'none';
        }
        if(orientation === 'landscape') {
          portraitContent.style.display = 'none';
          landscapeContent.style.display = 'block';
        }
        currentOrientation = orientation;
      }
      // register event
      switchButton.addEventListener('click', function() {
        if(currentOrientation === 'portrait') {
          return setUpOrientation('landscape')
        }
        if(currentOrientation === 'landscape') {
          return setUpOrientation('portrait')
        }
      })
      // on page load
      setUpOrientation('portrait')
    </script>
  </body>
</html>
```

#### Passed example 2

A page is always displayed in `landscape` mode, whatever the device orientation is. A switch button on the page can be activated to displayed the page according to the device orientation. (eg: alternative `CSS` is applied to make page operable).

A page where the content is adaptable based on orientation. In this case `CSS` is used to make page operable, by showing and hiding relevant content.

```html
<html lang="en">
  <head>
    <style>
      @media (orientation: portrait) {
        .show-me-portrait {
          display: block;
        }
        .show-me-landscape {
          display: none;
        }
      }
       @media (orientation: landscape) {
        .show-me-portrait {
          display: none;
        }
        .show-me-landscape {
          display: block;
        }
      }
    </style>
  </head>
  <body>
    <section class='show-me-portrait'>
        I am operable in PORTRAIT
    </section>
    <section class='show-me-landscape'>
        I am operable in LANDSCAPE
    </section>
  </body>
</html>
```

### Failed

#### Failed example 1

A page is always displayed in `landscape` mode, irrespective of the device orientation. A non-functional switch button is provided on the page.

```html
<html lang="en">
  <head>
  </head>
  <body>
    <button id='switch-button'>
      Switch Orientation does nothing :(
    </button>
    <section id='show-me-landscape'>
        I am only operable in LANDSCAPE
    </section>
    <script>
      var switchButton = document.getElementById('switch-button');
      var landscapeContent = document.getElementById('show-me-landscape');
    </script>
  </body>
</html>
```

#### Failed example 2

A page contains a functional visible and usable only in `landscape` mode that disappears in portrait mode. There is no functionality in the page to switch content based on orientation.

```html
<html lang="en">
  <head>
    <style>
      @media (orientation: portrait) {
        .show-me-landscape {
          transform: rotate(-90deg);
        }
      }
       @media (orientation: landscape) {
        .show-me-landscape {
          display: block;
        }
      }
    </style>
  </head>
  <body>
    <section class='show-me-landscape'>
        I am ALWAYS operable in LANDSCAPE
    </section>
  </body>
</html>
```

### Inapplicable

#### Inapplicable example 1

Content on the page can be viewed both in `landscape` and `portrait` mode.

```html
<html lang="en">
  <body>
    I am a simple page
  </body>
</html>
```

#### Inapplicable example 2

A locked display orientation is essential to the content on the page. A piano web application requires the device to be in landscape mode to allow room for enough of the piano keys to be functionally usable. 

```
Note: IMAGINE A PIANO APPLICATION, that is operable only in one orientation
```